Add ContextPropagation runtime util + captureContext operator (#3145)
reactor/reactor-core: docs/asciidoc/advanced-contextPropagation.adoc

[[context.propagation]]
= Context-Propagation Support

Since 3.5.0, Reactor-Core embeds support for the `io.micrometer:context-propagation` SPI.
This library is intended as a means to easily adapt between various implementations of the concept of a Context, of which
`ContextView`/`Context` is an example, and between `ThreadLocal` variables as well.

`ReactorContextAccessor` allows the Context-Propagation library to understand Reactor `Context` and `Contextview`.

> It implements the SPI and is loaded via `ServiceLoader`.

Reviewer: ```suggestion
It implements the SPI and is loaded via `java.util.ServiceLoader`.
```

> No user action is required, other than depending on reactor-core and `context-propagation` (the class is public but shouldn't generally be accessed by user code).

Reviewer: ```suggestion
No user action is required, other than having a dependency on both reactor-core and `io.micrometer:context-propagation`. The `ReactorContextAccessor` class is public but shouldn't generally be accessed by user code.
```

> On top of that, Reactor-Core 3.5.0 also introduces operators that transparently deal with `ContextSnapshot`s if the library is available at runtime, for users' convenience.

Reviewer 2: `operators` implies there are more than 1. Currently there is only one, correct? We can change the wording once we have more.

== `contextCapture` Operator

> This operator is a convenient alternative to `contextWrite` which performs a `ContextSnapshot#capture` and uses the SPI to translate that snapshot into the Reactor `Context`.

Reviewer: ```suggestion
This operator is a convenient alternative to `contextWrite` to perform a `ContextSnapshot#capture` and then to use that snapshot to populate the Reactor `Context`.
```

> This operator is a convenient alternative to `contextWrite` to perform a `ContextSnapshot#capture` and then to use that snapshot to populate the Reactor `Context`.

Reviewer 2: I'd suggest first explaining the intent and then additional context. It's easier to understand first what reactor does without having to dive into `context-propagation` API if it is not necessary (and I understand this API is serving such a purpose). So I'd start with saying that if there are thread locals at subscription time that you need in upstream operators in reactor context, use this operator, but first register in `ThreadLocalAccessor`. I think it's easier to digest starting with the intent and then elaborating on the reason / implementation motivations.

Author: missed that one. will reword that part a bit in the next commit.

> As a result, if there were any threadlocals during subscription phase they would now be stored into the `Context` and visible

Reviewer: ```suggestion
As a result, if there were any ThreadLocal values during subscription phase, for which there is a registered `ThreadLocalAccessor`, their values would now be stored in the Reactor `Context` and visible
```

at runtime in upstream operators.

====
[source,java]
----
//assuming TL is known to Context-Propagation as key TLKEY.
static final ThreadLocal<String> TL = new ThreadLocal<>();

//in the main thread, tl is set to "HELLO"
TL.set("HELLO");

Mono.deferContextual(ctx ->
  Mono.delay(Duration.ofSeconds(1))

>       //we're now in another thread, tl is not set

Reviewer 2: `tl` -> `TL` to be accurate

>       .map(v -> "delayed " + ctx.getOrDefault(TLKEY, "not found") + ", tl=" + TL.get())

Reviewer 2: How about

    "delayed ctx[" + TLKEY + "]=" + ctx.getOrDefault(TLKEY, "not found") + ", TL=" + TL.get()

For me at least it would be easier to understand the result.
